### DISCUSSION TITLE
More efficient synchronization?

### DISCUSSION THREAD

**Discussion by n0lavar** [David-Haim/concurrencpp#109] [General]
I am using manual executor to run async on my main thread and looks like mutex work take a lot of time
It would be nice if you add lock-free queue or something like this to improve performance
![image](https://user-images.githubusercontent.com/30406125/209852023-a2c66fc5-61cb-4fe9-8c4d-698eec5d97bf.png)

**Comment by David-Haim** [comment]
I'm not sure a lock free queue here will solve anything.
Lock free containers are very problematic, and they don't always provide better performance than traditional containers that are synchronized with a lock.

In the manual_executor case, many producers can enqueue tasks to run and many consumers can dequeue those tasks and run them, so we need an unbounded MPMC lock free queue. those unbounded mpmc lock free queues work by allocating every element separately and cas-ing those elements into the queue. so we're basically trading using a lock with allocating memory for every push operation. not sure it will give any performance gain on windows (with windows' crappy implementation of `malloc`/`HeapAlloc`).

Moreover, manual_executor gives the user the ability to wait until tasks are available or until a number of tasks are executed. if we remove the lock, then we also remove the underlying condition variable, and there is no way for us to wait for events any more. we will have to run the executor running methods in a busy loop or something to gain the same functionality.

Over all, I think that if applications run into executor related performance problems that are caused by high contention, then maybe the tasks that are given to those executors are too small to execute, and serving a longer-running tasks might help tremendously to reduce contention. 

Also, `executor::post` is more efficient than `executor::submit` (but you don't get a `result` in return), maybe you can switch some of the tasks to use that?

**Comment by n0lavar** [comment]
I use only `co_await` and `resume_on` in my code and I don't really want to limit users of my engine with "please use only long raged tasks"... Async execution in the end of each frame on main thread is really widely used pattern in gamedev
Maybe some tricks with r/w/rw mutexes could help?
Another option is separate executor with lock-free queue which can be used with custom and fast general purpose allocator

**Comment by n0lavar** [comment]
We can also move mutex and deque types to the traits so user can define types he want, for example stub mutex and lock free container

**Comment by ShirenY** [comment]
Also a game programmer here looking for a Uni-Task like solution for C++, I'm curious after two year what's your experience on this? Can you share some thoughts on this idea? @n0lavar

**Comment by n0lavar** [comment]
@ShirenY I haven't used this library in a production game, it's still my pet project, though not abandoned. But here's my experience:
 * Overall a handy library for async. If you don't need anything extraordinary, you just write code.
 * Simple and clear executors, you can make your own for main, audio, input/output threads.
 * Its own thread pool.

But there are some disadvantages:
* The biggest disadvantage is that there is no option to [return to the executor that was before the feature was called](https://github.com/David-Haim/concurrencpp/discussions/161). As I recall, I asked about this feature too, but the author said it would add overhead even if you don't use it. This is understandable, but very inconvenient. Without this feature, the code smells like C. It would be quite ideal to wrap this in a RAII class, but C++ itself can't do that =)
* As stated in the post, many things are handy, but not performant enough. So far I haven't gotten to a situation where it's critical, but if I do, I can always fix it myself and even create MR. Some ideas - still try lock-free container (the problem from this post) and try another mutex (for example, from folly).
* Own implementations of read/write and reqursive mutexes. I needed them in asset manager, where many threads can read and only some can write, when the required asset was not found. I had to implement it myself.
* The library uses exceptions. Many gamedev projects completely turn them off in the project, you can't do that here. 

Visually the “bad” part turned out bigger, but it's not, it's just that the bad stuff is memorable =). It's still a great library.

**Reply by ShirenY** [reply]
Thanks for the detailed information! Your experience has been really helpful for me in evaluating this approach.

In my case, I'm leaning more toward a main-thread coroutine system, similar to Unity’s. It makes handling game logic much more straightforward. Threading, on the other hand, can be risky for GPP, so I prefer to encapsulate that complexity in a separate toolset, like a thread pool, rather than exposing something like co_await RunInThreadPool(). It’s too easy to use—and that makes me a bit nervous. :)

Right now, I’m torn between implementing a simple coroutine system for my engine, or writing a single-threaded executor for concurrentcpp. As you mentioned, we want short coroutines to be extremely lightweight. Of course, concurrentcpp offers far more than my minimal approach, but that also comes with synchronization overhead.

On top of that, I still need utilities like co_await next_frame()/late_update()/physics_update() for gameplay features. I also really wish there were a simple way to stop a coroutine—something like task_handle.stop().

It’s a tough decision to make. :)
